File: Codigos/PUJ_ML_01/examples/python3/LinearModel_Fit_GradientDescent_01.py
## =========================================================================
## @author Leonardo Florez-Valencia (florez-l@javeriana.edu.co)
## =========================================================================

import numpy, sys
sys.path.insert( 0, '../../lib/python3' )

from PUJ_ML_01.lib.python3.PUJ.Model.Linear import Linear
from PUJ_ML_01.lib.python3.PUJ.Optimizer import *

## ----------
## -- Main --
## ----------

# Load data
data = numpy.loadtxt( open( sys.argv[ 1 ], 'rb' ), delimiter = ',' )

# Configure model
m = Linear( )
m.setParameters( [ 0 for i in range( data.shape[ 1 ] ) ] )
print( 'Initial model = ' + str( m ) )

# Configure cost
J = Linear.Cost( m, data[ : , 0 : -1 ], data[ : , -1 : ] )

# Analytical model
a = Linear( numpy.matrix( data ) )

# Debugger
## debugger = PUJ.Optimizer.Debug.Simple
debugger = PlotPolynomialCost( data[ : , 0 : -1 ], data[ : , -1 : ] )

# Fit using an optimization algorithm
opt = GradientDescent( J )
opt.setDebugFunction( debugger )
opt.setLearningRate( 1e-4 )
opt.setNumberOfIterations( 10000 )
opt.setNumberOfDebugIterations( 10 )
opt.Fit( )

# Show results
print( '===========================================' )
print( '= Iterations       :', opt.realIterations( ) )
print( '= Fitted model     :', m )
print( '= Analytical model :', a )
print( '===========================================' )

## eof - $RCSfile$

File: PUJ_ML-3/examples/cxx/Linear_Model_Fit_GradientDescent_01.cxx
// =========================================================================
// @author Leonardo Florez-Valencia (florez-l@javeriana.edu.co)
// =========================================================================

#include <iostream>
#include <random>
#include <PUJ_ML/Model/Linear.h>
#include <PUJ_ML/Optimizer/GradientDescent.h>

using TScalar = double;
using TModel  = PUJ_ML::Model::Linear< TScalar >;
using TOptimizer = PUJ_ML::Optimizer::GradientDescent< TScalar >;

int main( int argc, char** argv )
{
  // Some parameters to generate sample data
  unsigned long long M = 10;
  TScalar minX = -100;
  TScalar maxX = 100;

  // Some model
  TModel dm( { 0.3, 2, 0.5 } );
  unsigned long long N = dm.numberOfParameters( );

  // Some random data
  std::random_device rd;
  std::mt19937 gen( rd( ) );
  std::uniform_real_distribution< TScalar > dis( minX, maxX );
  TModel::TMatrix X =
    TModel::TMatrix::Zero( M, dm.inputSize( ) ).unaryExpr(
      [&]( const TScalar& z ) -> TScalar { return( dis( gen ) ); }
      );
  auto Y = dm.evaluate( X );

  // Model to be optimized
  TModel om( std::vector< TScalar >( N, 0 ) );

  // Cost
  TModel::Cost cost( &om, X, Y );

  // Optimizer
  auto opt = TOptimizer( &cost );
  opt.setLearningRate( 1e-4 );
  opt.Fit( );

  // Show results
  std::cout << "******************************" << std::endl;
  std::cout << "* Original model  : " << dm << std::endl;
  std::cout << "* Optimized model : " << om << std::endl;
  std::cout << "******************************" << std::endl;

  return( EXIT_SUCCESS );
}

// eof - $RCSfile$

